cmd/vm_firewall.go

```diff
@@ -0,0 +1,166 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/exoscale/cli/table"
+	"github.com/exoscale/egoscale"
+	"github.com/spf13/cobra"
+)
+
+// vmFirewallCmd represents the vm firewall command
+var vmFirewallCmd = &cobra.Command{
+	Use:   "firewall",
+	Short: "Virtual machines firewall management",
+}
+
+// vmFirewallSetCmd represents the firewall set command
+var vmFirewallSetCmd = &cobra.Command{
+	Use:   "set <vm name> <SG name> [SG name] ...",
+	Short: "set the security groups of a virtual machine",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) < 2 {
+			return cmd.Usage()
+		}
+
+		sgs, err := getSecurityGroupsByNameOrID(args[1:])
+		if err != nil {
+			return err
+		}
+
+		vm, err := getVirtualMachineByNameOrID(args[0])
+		if err != nil {
+			return err
+		}
+
+		return setVirtualMachineSecurityGroups(vm, sgs)
+	},
+}
+
+// vmFirewallAddCmd represents the firewall add command
+var vmFirewallAddCmd = &cobra.Command{
+	Use:   "add <vm name> <SG name> [SG name] ...",
+	Short: "add security groups to a virtual machine",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) < 2 {
+			return cmd.Usage()
+		}
+
+		sgs, err := getSecurityGroupsByNameOrID(args[1:])
+		if err != nil {
+			return err
+		}
+
+		vm, err := getVirtualMachineByNameOrID(args[0])
+		if err != nil {
+			return err
+		}
+
+		sgToAdd := make([]egoscale.SecurityGroup, 0)
+
+	next:
+		// Check if requested security groups are not already set to the VM instance
+		for i := range sgs {
+			for j := range vm.SecurityGroup {
+				if sgs[i].ID.Equal(*vm.SecurityGroup[j].ID) {
+					continue next
+				}
+			}
+			sgToAdd = append(sgToAdd, sgs[i])
+		}
+
+		return setVirtualMachineSecurityGroups(vm, append(vm.SecurityGroup, sgToAdd...))
+	},
+}
+
+// vmFirewallRemoveCmd represents the firewall remove command
+var vmFirewallRemoveCmd = &cobra.Command{
+	Use:   "remove <vm name> <SG name> [SG name] ...",
+	Short: "remove security groups from a virtual machine",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) < 2 {
+			return cmd.Usage()
+		}
+
+		sgs, err := getSecurityGroupsByNameOrID(args[1:])
+		if err != nil {
+			return err
+		}
+
+		vm, err := getVirtualMachineByNameOrID(args[0])
+		if err != nil {
+			return err
+		}
+
+		sgRemaining := make([]egoscale.SecurityGroup, 0)
+
+	next:
+		for i := range vm.SecurityGroup {
+			for j := range sgs {
+				if sgs[j].ID.Equal(*vm.SecurityGroup[i].ID) {
+					continue next
+				}
+			}
+			sgRemaining = append(sgRemaining, vm.SecurityGroup[i])
+		}
+
+		return setVirtualMachineSecurityGroups(vm, sgRemaining)
+	},
+}
+
+// setVirtualMachineSecurityGroups sets a virtual machine instance security groups.
```

Comment:
This call does "set" and "show", I'd split the two actions even if it means doing more API calls.

Reply:
Fixed in 057d37037d25fb3755ced13b28a6f24277864868.